Source/Task/AsyncLib.cpp

```diff
@@ -265,8 +257,46 @@ class AsyncBlockInternalGuard
     }
 
 private:
-    AsyncBlockInternal * const m_internal1;
-    AsyncBlockInternal * m_internal2;
+    AsyncBlockInternal * const m_internal;
+    AsyncBlockInternal * m_userInternal;
+
+    // Locks the correct async block and returns a pointer to the one
+    // we locked.
+    static AsyncBlockInternal* DoLock(_In_ AsyncBlock* asyncBlock)
+    {
+        AsyncBlockInternal* internal = reinterpret_cast<AsyncBlockInternal*>(asyncBlock->internal);
+        ASSERT(internal);
+        while (internal->lock.test_and_set()) {}
+
+        // We've locked the async block. We only ever want to keep a lock on one block
+        // to prevent deadlocks caused by lock ordering.  If the state is still valid
+        // on this block, we ensure the async block we're locking is the permanent one
+        // associated with the async state.  
+
+        if (internal->state != nullptr && asyncBlock != &internal->state->asyncBlock)
+        {
+            // Grab a state ref here because releasing the lock can allow
+            // the state to be cleared / released.
+            AsyncStateRef state(internal->state);
+            internal->lock.clear();
+            internal = DoLock(&state->asyncBlock);
+
+            // We locked the right object, but we need to check here to see if we
```

Comment:
I'm confused by this part.

When checking for `internal->state == nullptr`, are we checking if we are racing with someone else?
The heavy reuse of the local variable `internal` with different values makes it hard to follow what's happening. Can you use different names? (I know it inhibits NRVO, but it's a pointer so I don't think NRVO applies anyway)

Reply:
Yup, I can clarify.  Once we have internal assigned, we've been granted the lock so there is no chance of another actor messing with state.